source/mir/math/stat.d

```diff
@@ -461,6 +444,423 @@ unittest
     assert(float.max.repeat(3).hmean!(double, "fast").approxEqual(float.max));
 }
 
+private
+F nthroot(F)(in F x, in size_t n)
+    if (isFloatingPoint!F)
+{
+    import mir.math.common: sqrt, pow;
+
+    assert(x > 0, "nthroot: Can only take nth root of positive numbers");
+
+    if (n > 2) {
+        return pow(x, cast(F) 1 / cast(F) n);
+    } else if (n == 2) {
+        return sqrt(x);
+    } else if (n == 1) {
+        return x;
+    } else {
+        return cast(F) 1;
+    }
+}
+
+version(mir_test)
+@safe @nogc pure nothrow
+unittest {
+    import mir.math.common: approxEqual;
+
+    assert(nthroot(9.0, 0).approxEqual(1));
+    assert(nthroot(9.0, 1).approxEqual(9));
+    assert(nthroot(9.0, 2).approxEqual(3));
+    assert(nthroot(9.5, 2).approxEqual(3.08220700));
+    assert(nthroot(9.0, 3).approxEqual(2.08008382));
+}
+
+/++
+Output range for gmean.
++/
+struct GMeanAccumulator(T, ProdAlgo prodAlgo) 
+    if (isMutable!T && isFloatingPoint!T)
+{
+    import mir.math.numeric: ProdAccumulator;
+
+    ///
+    size_t count;
+    ///
+    ProdAccumulator!(T, prodAlgo) prodAccumulator;
+
+    ///
+    F gmean(F = T)() @property
+        if (isFloatingPoint!F)
+    {
+        return nthroot(cast(F) prodAccumulator.prod, count);
```

Comment:
`prodAccumulator.prod` may have exponent overflow (in case of separate exponent accumulation), while the `gmean` itself can be defined.

In the case of the exponent overflow, we can define `gmean` as
```d
nthroot(cast(F) prodAccumulator.mantissa, count) * exp2(cast(F) prodAccumulator.exp / count)
```

Reply:
I need to add some stuff to `prod` or disable the `naive` algorithm for `gmean`. I will not have time to work on this until later today most likely.

Reply:
We can remove the algorithm selection completely from ProdAccumulator and use only floating-point accumulation.

Later we can replace the hardware floating-point accumulation with extended software floating-point accumulation. I am working for it as part of Ion,
https://github.com/libmir/mir-ion/blob/master/source/mir/bignum/fp.d

It would make API simpler and the precision would be improved a lot.

Reply:
The fix I was thinking about for `ProdAccumulator` was actually pretty simple. I was going to add `exp` and `manitssa` member functions that would also work when the algorithm is `naive` and `isFloatingPoint!T` is true. Since `nthroot` requires the type to be floating point, then it would not be an issue.

I have no issue with future changes to the algorithm so that it uses what you are doing for `mir.ion` on the floating point side of things. However, `naive` was added so that `prod` can handle complex types and user-defined types. Removing that would eliminate the ability to call `prod` for those types, which some people may want. It also makes it easy to implement a `prod` function that only takes integer types and returns an integer type (ideally this would have its own implementation, but I considered that something for future work). Regardless, I think that has some uses. Precision isn't the concern for integer types, overflow is. And, where overflow is a concern, people should be using `long`s or `std.bigint`.

Reply:
We need Prod only for statistics. User-defined and complex types can use `reduce` for product and they can't use `gmean` anyway.

Reply:
@9il Ok, I should have something later today or tomorrow fixing this. 

It occurs to me that one of my issues with `reduce` is the same as the reason why phobos has both `std.algorithm.iteration.reduce` and `std.algorithm.iteration.fold`. `reduce` is not easily used in UFCS chains because the slice/array/range is the second parameter. 

Perhaps I can add a separate `fold` function to mir.algorithm?

Reply:
> Perhaps I can add a separate fold function to mir.algorithm?

LGTM